Add option to configure Tracks to store its DB in Application Support instead of Documents

mokagio opened the pull request:
See #146.

~~This is just a test to verify my internal non-sandboxed macOS app works fine when using Application Support.~~

Update: we decided to merge this change, but in a way that wouldn't change the behavior of Simplenote macOS.

To test these changes:

- I added unit tests 🙂, although CI is failing 🙃, but not because of my tests 🙂
- Checkout this PR in Simplenote macOS https://github.com/Automattic/simplenote-macos/pull/683

mokagio committed:
Add logic to get URL for Application Support folder

mokagio committed:
Store Tracks.sqlite store in Application Support instead of Documents

mokagio committed:
Add test for TracksContexManager store URL location

mokagio committed:
Refactor: extract context manager store URL definition

This will allow us to add the conditional logic for the sandbox vs
non-sandbox location later on.

mokagio committed:
Add nullability annotation to TracksContextManager properties

We'll need this later to make the custom init with the sandbox mode flag
non-nullable.

mokagio committed:
Allow TracksContextManager consumers to configure the store location

mokagio commented on Automattic-Tracks-iOS/Model/Core Data/TracksContextManager.h:
```diff
@@ -3,8 +3,22 @@
-@property (nonatomic, strong) NSManagedObjectContext *managedObjectContext;
-@property (nonatomic, strong) NSManagedObjectModel *managedObjectModel;
-@property (nonatomic, strong) NSPersistentStoreCoordinator *persistentStoreCoordinator;
+@property (nonnull, nonatomic, strong) NSManagedObjectContext *managedObjectContext;
+@property (nonnull, nonatomic, strong) NSManagedObjectModel *managedObjectModel;
+@property (nonnull, nonatomic, strong) NSPersistentStoreCoordinator *persistentStoreCoordinator;
```
I had to add the annotation here too to avoid warnings. It makes the diff a bit noisier but it's good to have it anyways.

The implementation of all those getters makes sure they are not `nil`, or rather, fails if they are, so `nonnullable` seemed appropriate.

jleandroperez replied:
Yup that's the perfect approach Gio!

mokagio commented on Automattic-Tracks-iOS/Model/Core Data/TracksContextManager.h:
```diff
@@ -3,8 +3,22 @@
+/// the app is running and where to save the store.
+///
+/// The `super` `init` method defaults to `false` for backward compatibility.
+- (nonnull instancetype)initWithSandboxedMode:(BOOL)sandboxed NS_DESIGNATED_INITIALIZER;
```
I wonder whether the parameter should be an enum instead:

```swift
_ = TracksContextManager(storeLocation: .applicationSupport)
// or
_ = TracksContextManager(storeLocation: .documents)
```

I'm on the fence and might sit overnight. Regardless, because the changes doesn't affect existent consumers thanks to the `init` override below, I think it's okay to merge this as it is and then I can follow up with a refactor if we think it's worth it.

jleandroperez replied:
To be **entirely** honest, I would keep the initializer simple, and just switch to the Application Support folder.

Something in the lines of:

- If a `Documents` folder already exists, use it
- Otherwise, check if a folder in `Applications Support` exists, and use it
- If none of the above, create a new folder in `Applications Support`

WYDT?

jkmassel replied:
I'm ok with either approach. I like how pretty the enum is, but I like how much work @jleandroperez's isn't 😄

mokagio commented on Automattic-Tracks-iOS/Model/Core Data/TracksContextManager.m:
```diff
@@ -3,10 +3,25 @@
+
+- (instancetype)initWithSandboxedMode:(BOOL)sandboxed {
+    self = [super init];
+    if (!self) { return nil; }
```
This should never happen so I defined the initializer as `nonnullable`. In Swift land, that would be the case.

jleandroperez replied:
Standard ObjC, exactly!

mokagio reviewed (commented):
I refined the implementation to avoid changing the way existing consumers of the library store the DB.

@jleandroperez @jkmassel can I ask you for a new round of review?

jleandroperez commented on Automattic-Tracks-iOS/Model/Core Data/TracksContextManager.m:
```diff
@@ -3,10 +3,25 @@
 @implementation TracksContextManager
 
+- (instancetype)init {
+    return [[TracksContextManager alloc] initWithSandboxedMode:true];
```
This initializer returning a new instance, rather than the one being initialized.

You would typically instantiate a Context Manager like this:

```
TracksContextManager *manager = [[TracksContextManager alloc] init];
```

1. `alloc` reserves memory for the object
2. `init` is expected to initialize the instance

In this case, `init` is alloc'ating more memory, and returning yet another instance.

mokagio replied:
Thank you for this! Gosh 😳 I'm really rusty at writing Objective-C 🤦 

I felt like there was something wrong with that, but it compiled and the tests passed so I shrugged it off.

Interestingly, Xcode also gives me a warning about it

![image](https://user-images.githubusercontent.com/1218433/94876763-fca44c80-049b-11eb-9fde-d0c732f77a53.png)

It either came after I wrote the code and moved on or after I added the `NS_DESIGNATED_INITIALIZER` annotation. Regardless, I shouldn't have missed it... 😳

jleandroperez commented on Automattic-Tracks-iOS/Model/Core Data/TracksContextManager.m:
```diff
@@ -3,10 +3,25 @@
 @implementation TracksContextManager
 
+- (instancetype)init {
+    return [[TracksContextManager alloc] initWithSandboxedMode:true];
```
```suggestion
    return [self initWithSandboxedMode:true];
```

jleandroperez replied:
(Sending you a change suggestion to address that!)

jleandroperez commented on Automattic-Tracks-iOS/Model/Core Data/TracksContextManager.m:
```diff
@@ -59,6 +74,64 @@ - (NSPersistentStoreCoordinator *)persistentStoreCoordinator {
+        abort();
+    }
+
+    return applicationSupportURL;
```
You can proooobably simplify this API as follows:

```
NSURL *appSupportURL = [[[NSFileManager defaultManager] URLsForDirectory:NSApplicationSupportDirectory inDomains:NSUserDomainMask] lastObject];
```

[Ref. here](https://github.com/Simperium/simperium-ios/blob/develop/Simperium/SPPersistentMutableDictionary.m#L400)

(The Application Support path is expected to always be there!)

jleandroperez reviewed (changes requested):
@mokagio sir!! Few changes suggested:

- Initializer fix (would prevent leaking memory)
- ApplicationSupport API simplification (not required, not a blocker either!!)

As per the initializer detail, 10000% up to you!

- Taking that parameter isn't a problem at all
- It can be solved without specifying if we're sandboxed or not, as well. Yet another alternative, is to just always use the Application Support path, and [migrate legacy databases, if needed](https://github.com/Simperium/simperium-ios/blob/develop/Simperium/SPPersistentMutableDictionary.m#L374)

I would take the easiest path. As usual 😄

mokagio committed:
Call self in TracksContextManager convenience initializer

Thanks @jleandroperez!

See https://github.com/Automattic/Automattic-Tracks-iOS/pull/147#discussion_r498425403

mokagio committed:
Simplify code to get URL of Application Support folder

Thanks @jleandroperez

See https://github.com/Automattic/Automattic-Tracks-iOS/pull/147#discussion_r498429821

mokagio commented:
Thanks @jleandroperez for the great review!

Regarding the init, I like the idea of keeping it simple and eventually migrate everything to Application Support.

I haven't tried, but I think that even to just check whether a file exists in the Documents folder from a non-sandboxed app, macOS would request the user for permission. If that's the case, then we'd be back to square one, because the real aim of this change is avoid that behavior in non-sandboxed apps.

I really like the migration idea, but I think it needs more experimenting and might be best for a dedicated PR.

jleandroperez commented on Automattic-Tracks-iOS/Model/Core Data/TracksContextManager.m:
```diff
@@ -59,6 +74,51 @@ - (NSPersistentStoreCoordinator *)persistentStoreCoordinator {
+        return [self applicationSupportURLForContainerApp];
+    } else {
+        return [self applicationDocumentsDirectory];
+    }
```
@mokagio Is it possible this logic is inverted? (sandboxed > Documents Directory?).

Otherwise non sandboxed apps would still trigger the Documents Folder alert perhaps?

mokagio replied:
You are totally right!

I didn't notice because the tests were inverted, too 🤦‍♂️

jleandroperez reviewed (approved):
@mokagio One tiny comment sent away (apologies I missed that!!).

Other than that, please :shipit: when ready!!

mokagio committed:
Fix sandboxed behavior tests expectation: sandboxed goes to Documents

This should have always been the case, but somehow I got confused and
didn't realize my mistake.

Thanks @jleandroperez for picking this up! Goes to show that tests are
only good when written properly!

https://github.com/Automattic/Automattic-Tracks-iOS/pull/147/files/3e38df8087e7b38a76b45acc5e89527746b4145b#r498830462

mokagio commented:
😫 I feel like I'm in a deadlock... There's some inconsistent issue with the tests, which [doesn't happen with the Xcode 12](https://github.com/Automattic/Automattic-Tracks-iOS/pull/150), but I can't upgrade to Xcode 12 because [the pod validation fails](https://github.com/Automattic/Automattic-Tracks-iOS/issues/149).

Inconsistent as in [this](https://app.circleci.com/pipelines/github/Automattic/Automattic-Tracks-iOS?branch=store-data-in-application-support):

![image](https://user-images.githubusercontent.com/1218433/95062472-baebfe00-0748-11eb-8804-deece0a1b612.png)

jleandroperez reviewed (approved):
:shipit: sir!!

jleandroperez commented:
For the record, Cocoapods had issues finding the Sodium-Fork framework. Workaround involved running `bundle exec pod repo update`

Thank you Gio!!
